CMakeLists.txt

```diff
@@ -0,0 +1,232 @@
+
+    set(BIN_FILE $<CONFIG>/${EXECUTABLE}${EXECUTABLE_VERSION_SUFFIX}.bin)
+    set(HEX_FILE $<CONFIG>/${EXECUTABLE}${EXECUTABLE_VERSION_SUFFIX}.hex)
+    set(MAP_FILE $<CONFIG>/${EXECUTABLE}${EXECUTABLE_VERSION_SUFFIX}.map)
+
+    # add link options to generate memory map
+    target_link_libraries(${EXECUTABLE} -Wl,\"-Map=${MAP_FILE}\",--cref)
+
+    # generate .bin firmware file
+    add_custom_command(TARGET ${EXECUTABLE}
+        POST_BUILD
+        COMMAND ${CMAKE_OBJCOPY} -O binary $<TARGET_FILE:${EXECUTABLE}> ${BIN_FILE}
+        WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+        COMMENT "Objcopying ${EXECUTABLE} to firmware ${BIN_FILE}"
+        DEPENDS ${EXECUTABLE}
+        BYPRODUCTS ${BIN_FILE})
+
+    # generate .hex firmware file
+    add_custom_command(TARGET ${EXECUTABLE}
+        POST_BUILD
+        COMMAND ${CMAKE_OBJCOPY} -O ihex $<TARGET_FILE:${EXECUTABLE}> ${HEX_FILE}
+        WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+        COMMENT "Objcopying ${EXECUTABLE} to firmware ${HEX_FILE}"
+        DEPENDS ${EXECUTABLE}
+        BYPRODUCTS ${HEX_FILE})
+
+    # Clean up artefacts
+    set_property(TARGET ${EXECUTABLE}
+        PROPERTY
+            ADDITIONAL_CLEAN_FILES ${BIN_FILE} ${MAP_FILE} ${HEX_FILE})
+
+    # find_program(PRETTYSIZE_CMD prettysize prettysize.py)
+
+    # if(NOT PRETTYSIZE_CMD STREQUAL "PRETTYSIZE_CMD-NOTFOUND")
+    # add_custom_command(
+    # TARGET ${EXECUTABLE} POST_BUILD
+    # COMMAND prettysize
+    # -l ${CMAKE_CURRENT_LIST_DIR}/linker_script_rz_a1l.ld
+    # -s ${CMAKE_SIZE} $<TARGET_FILE:${EXECUTABLE}>
+    # WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+    # COMMENT "Displaying usage graph for ${EXECUTABLE}")
+    # endif()
+endfunction(add_firmware)
+
+# Global compilation options
+add_compile_options(
+
+    # Set preinclude
+    $<$<COMPILE_LANGUAGE:CXX>:-includepreinclude_cxx.h>
+
+    # Feature flags
+    -fdiagnostics-parseable-fixits
+    -fsigned-char
+
+    # Debug symbols
+    $<$<CONFIG:DEBUG>:-g> # Include
+    $<$<CONFIG:DEBUG>:-ggdb> # Include
+    $<$<CONFIG:RELEASE>:-s> # Strip
+
+    # Optimization level
+    $<$<CONFIG:DEBUG>:-Og>
+    $<$<CONFIG:RELEASE>:-O2>
+
+    # Link time optimizations
+    $<$<CONFIG:RELEASE>:-flto>
+    -ffunction-sections
+    -fdata-sections
+
+    # C++ stuff
+    $<$<COMPILE_LANGUAGE:CXX>:-fno-rtti>
+    $<$<COMPILE_LANGUAGE:CXX>:-fno-exceptions>
+    $<$<COMPILE_LANGUAGE:CXX>:-fno-use-cxa-atexit>
+    $<$<COMPILE_LANGUAGE:CXX>:-fno-threadsafe-statics>
+
+    # ASM stuff
+    $<$<COMPILE_LANGUAGE:ASM>:-x>
+    $<$<COMPILE_LANGUAGE:ASM>:assembler-with-cpp>
+
+    # Warnings
+    # $<$<CONFIG:DEBUG>:-Wextra> # TODO: Enable when ready
+    # $<$<CONFIG:DEBUG>:-Wall> # TODO: Enable when ready
+    # $<$<CONFIG:DEBUG>:-Wpedantic> # TODO: Enable when ready
+    $<$<CONFIG:DEBUG>:-Wnull-dereference>
+    $<$<CONFIG:DEBUG>:-Wno-psabi>
+
+    # Stack usage
+    $<$<CONFIG:DEBUG>:-Wstack-usage=100>
+    $<$<CONFIG:RELEASE>:-Wstack-usage=1000>
```

Comment:
is this the new hotness on how to conditionally add flags vs. `if (){}` and appending to the language/build specific flag lists?

Reply:
lol yes. Generator expressions are 10 years old (CMake 3.0) but have only gotten more powerful over time